### ISSUE TITLE
Performance issue when using GeoJsonLayer

### ISSUE THREAD

**Issue by jfsiii** [visgl/deck.gl#1067]
I'm using `deck.gl` & `react-map-gl` in my app and am seeing the main thread get blocked for ~3 seconds when using a `GeoJsonLayer`. I made an example ([map](https://silver-saxophone.glitch.me/), [code](https://glitch.com/edit/#!/silver-saxophone?path=client/App.js)) based off the [GeoJSON example in the deck.gl repo](https://github.com/uber/deck.gl/blob/master/examples/geojson/app.js) which shows the behavior/

If you [look at the timeline where I loaded & refreshed 3 times (2 normal; 1 hard)](https://chromedevtools.github.io/timeline-viewer/?loadTimelineFromURL=https://cdn.glitch.com/38402265-c018-45c8-a96f-a1626887e9eb%2Fabout-blank-to-silver-saxophone.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508781329730), you can see the ~3 sec pauses on `getProgramParameter`
![this timeline](https://cdn.glitch.com/38402265-c018-45c8-a96f-a1626887e9eb%2Fsilver-saxophone.glitch.me-timeline.png?1508781168506)

Since this looks like it's from WebGL or one of `deck.gl`, `react-map-gl` or `mapbox-gl-js`, I wasn't sure how to improve performance by changes to the app code, but I realized [the GeoJSON file](https://gist.githubusercontent.com/jfsiii/b031dbbe7385d2c6822468a001d62225/raw/8ad146ac7db178b82480b8eb46eaaf406f60dc5d/USA-boundaries.json) is pretty large (~3MB) so I simplified the file down to ~30K. However, that just moved the ~3 sec pauses to a different function ([timeline](https://chromedevtools.github.io/timeline-viewer/?loadTimelineFromURL=https://cdn.glitch.com/eb8c029b-6783-4bcd-87fe-df6b46584e00%2Fabout-blank-to-reminiscent-glue.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508781972493)) ![other timeline](https://cdn.glitch.com/eb8c029b-6783-4bcd-87fe-df6b46584e00%2Freminiscent-glue.glitch.me-timeline.png?1508781957386)

I made an example with the 30K file as well ([map](https://reminiscent-glue.glitch.me), [code](https://glitch.com/edit/#!/reminiscent-glue?path=client/App.js:1:0))

I'm not sure if the issue lies with my code (but I'm following the example), or code in one of`deck.gl`, `react-map-gl` or `mapbox-gl-js`, but I wanted to file this to get the discussion started.

Please let me know if I should file this in another repo, or if there's anything else I can do.

_Note_: These traces were taken in Chrome Stable - Version 62.0.3202.62 (Official Build) (64-bit) in an Incognito Window. The results were the same in Chrome 61. I'm on macOS 10.12.6 (16G29)

**Comment by ibgreen** [comment]
@jfsiii Thanks for report and the detailed traces, extremely helpful! 

We very recently started to see a similar 0.5-3s delay in another app creating a simple `GeoJson`, so this definitely is on our radar, and your findings are interesting.
 
* The "hint" I get from your traces that this could be related to "WebGL readback" somehow stalling the GPU pipeline, either in mapbox or deck. 

* As far as I can tell, the `getProgramParameter` calls are mapbox-gl stack traces, not deck.gl stack traces. But per your report, the perf issue seems triggered by the combination of the two!

* The second call is less conclusive. I am a little unsure of where it comes from. If from deck/luma it could be related to picking or logging, but the stack should be deeper... It could also be from mapbox-gl. Is there anyway for you to see more information?

I quickly ran your example on Chrome 58 and Chrome Canary 64 on the latest 2016 MacBook and could not reproduce the delays. But given that we have seen something similar I have no doubts that what you see is real and needs to be addressed.

@georgios-uber @shaojingli @1chandu

**Comment by jfsiii** [comment]
@ibgreen Thanks for the quick response.

I'm glad to help, but I'm not sure what you mean by "Is there anyway for you to see more information?". Can you give me some more detail?

I've published examples which you can load in your browser and debug with your tools. You can fork ("remix") the project and edit it or download it to run locally

<img width="143" alt="screen shot 2017-10-23 at 2 23 25 pm" src="https://user-images.githubusercontent.com/57655/31914008-ebb0353a-b7fd-11e7-9748-8cb1b96c39fc.png"><img width="138" alt="screen shot 2017-10-23 at 2 23 40 pm" src="https://user-images.githubusercontent.com/57655/31914007-eb99956e-b7fd-11e7-9c97-46b18ef955b5.png">

You can also view the traces I took directly in your browser or download them and import them in Chrome's Performance tab
<img width="424" alt="screen shot 2017-10-23 at 2 28 21 pm" src="https://user-images.githubusercontent.com/57655/31914164-7d0a2e46-b7fe-11e7-8c1e-3fd837ba4b2f.png">

I re-ran the tests in Chrome 64 but the results were largely the same. However, I do have a 27" 4K external monitor, which seems relevant :) I run it scaled to 2560x1440. I re-ran the tests with no scaling in Chrome 62 & Chrome 64. The 30K example seems unchanged (still ~3s pauses at `getImageData`) but the behavior 3MB example changed. Now there were 3 different significant 
pauses

1. 800ms at `checkStatus`
1. 2650ms at `Program` in `Painter`
1. ~1100ms at `Texture.update`


Here's a list of examples to test and results from various runs on my local setup. 

My changes to browser, screen resolution seem to have some effect on the 3MB JSON example, but nothing seems to affect the behavior of the 30K JSON example.

### 3MB GeoJSON - pauses on `getProgramParameter`
- [demo](https://silver-saxophone.glitch.me/)
- [code](https://glitch.com/edit/#!/silver-saxophone?path=client/App.js)
- timeline - Chrome 62 - scaled [view](https://chromedevtools.github.io/timeline-viewer/?loadTimelineFromURL=https://cdn.glitch.com/38402265-c018-45c8-a96f-a1626887e9eb%2Fabout-blank-to-silver-saxophone.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508781329730) [download](https://cdn.glitch.com/38402265-c018-45c8-a96f-a1626887e9eb%2Fabout-blank-to-silver-saxophone.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508781329730)
- timeline - Chrome 64 - scaled [view](https://chromedevtools.github.io/timeline-viewer/?loadTimelineFromURL=https://cdn.glitch.com/38402265-c018-45c8-a96f-a1626887e9eb%2FChrome-64-about-blank-to-silver-saxophone.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508791993145) [download](https://cdn.glitch.com/38402265-c018-45c8-a96f-a1626887e9eb%2FChrome-64-about-blank-to-silver-saxophone.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508791993145)
- timeline - Chrome 62 - not scaled [view](https://chromedevtools.github.io/timeline-viewer/?loadTimelineFromURL=https://cdn.glitch.com/38402265-c018-45c8-a96f-a1626887e9eb%2FChrome-62-not-scaled-about-blank-to-silver-saxophone.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508795858862) [download](https://cdn.glitch.com/38402265-c018-45c8-a96f-a1626887e9eb%2FChrome-62-not-scaled-about-blank-to-silver-saxophone.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508795858862
)
- timeline - Chrome 64 - not scaled [view](https://chromedevtools.github.io/timeline-viewer/?loadTimelineFromURL=https://cdn.glitch.com/38402265-c018-45c8-a96f-a1626887e9eb%2FChrome-64-not-scaled-about-blank-to-silver-saxophone.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508794822576) [download](https://cdn.glitch.com/38402265-c018-45c8-a96f-a1626887e9eb%2FChrome-64-not-scaled-about-blank-to-silver-saxophone.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508794822576)

### 30K GeoJSON - pauses on `getImageData`
- [demo](https://reminiscent-glue.glitch.me/)
- [code](https://glitch.com/edit/#!/reminiscent-glue?path=client/App.js)
- timeline - Chrome 62 - scaled [view](https://chromedevtools.github.io/timeline-viewer/?loadTimelineFromURL=https://cdn.glitch.com/eb8c029b-6783-4bcd-87fe-df6b46584e00%2Fabout-blank-to-reminiscent-glue.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508781972493) [download](https://cdn.glitch.com/eb8c029b-6783-4bcd-87fe-df6b46584e00%2Fabout-blank-to-reminiscent-glue.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508781972493)
- timeline - Chrome 64 - scaled [view](https://chromedevtools.github.io/timeline-viewer/?loadTimelineFromURL=https://cdn.glitch.com/eb8c029b-6783-4bcd-87fe-df6b46584e00%2FChrome-64-about-blank-to-reminiscent-glue.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508792076274) [download](https://cdn.glitch.com/eb8c029b-6783-4bcd-87fe-df6b46584e00%2FChrome-64-about-blank-to-reminiscent-glue.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508792076274)
- timeline - Chrome 62 - not scaled [view](https://chromedevtools.github.io/timeline-viewer/?loadTimelineFromURL=https://cdn.glitch.com/eb8c029b-6783-4bcd-87fe-df6b46584e00%2FChrome-62-not-scaled-about-blank-to-reminiscent-glue.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508796821736) [download](https://cdn.glitch.com/eb8c029b-6783-4bcd-87fe-df6b46584e00%2FChrome-62-not-scaled-about-blank-to-reminiscent-glue.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508796821736)
- timeline - Chrome 64 - not scaled [view](https://chromedevtools.github.io/timeline-viewer/?loadTimelineFromURL=https://cdn.glitch.com/eb8c029b-6783-4bcd-87fe-df6b46584e00%2FChrome-64-not-scaled-about-blank-to-reminiscent-glue.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508794699547) [download](https://cdn.glitch.com/eb8c029b-6783-4bcd-87fe-df6b46584e00%2FChrome-64-not-scaled-about-blank-to-reminiscent-glue.glitch.me-soft-refresh-x2-hard-refresh-20171023T103321.json?1508794699547)

Here's some more information about my hardware:

![screen shot 2017-10-23 at 1 54 52 pm](https://user-images.githubusercontent.com/57655/31914312-0a4c3e70-b7ff-11e7-9d0d-9cfbbe7ca166.png)
![screen shot 2017-10-23 at 1 54 58 pm](https://user-images.githubusercontent.com/57655/31914311-0a34ad32-b7ff-11e7-9a7c-1f76e8a5a2ca.png)
<img width="668" alt="screen shot 2017-10-23 at 1 55 38 pm" src="https://user-images.githubusercontent.com/57655/31914310-0a0fb112-b7ff-11e7-9b9f-1969b2552b2d.png">

**Comment by ibgreen** [comment]
> I've published examples which you can load in your browser and debug with your tools.

Yes thanks and it is extremely helpful. I have done an initial attempt but not able to reproduce the lag. Will need to try again on an older computer with different Chrome versions.

> Is there anyway for you to see more information?". Can you give me some more detail?

Yes, I was just trying to understand if this was from mapbox or deck.gl. Since the functions calling `getImageData` are anonymous, it is hard to tell for sure, without e.g. a file name but I suppose that file names and other source map related stuff are not available in the profiler. Sorry for being brief. 

My hunch is that I don't think we have such code in deck.gl so I suspect it comes from mapbox in both cases. But still the perf issue seems triggered by deck...

Absolutely not my intention to push more work your way, more as a note to self: there is a possibly related thread (Suspected Chrome changes to webgl canvas interaction, affecting deck and mapbox integration), I think it could be worth testing the approach of adding a third transparent canvas on top of the mapbox and deck canvases and reading out from it once #988.

**Comment by jfsiii** [comment]
> Yes, I was just trying to understand if this was from mapbox or deck.gl. Since the functions calling getImageData are anonymous, it is hard to tell for sure, without e.g. a file name but I suppose that file names and other source map related stuff are not available in the profiler. Sorry for being brief.

No problem. I forgot the sourcemaps were inaccurate. I've been on quite the yak shaving safari with this. Two other demo sites had threw errors when loading `react-map-gl` (https://github.com/stackblitz/core/issues/173 & https://github.com/CompuIves/codesandbox-client/issues/250), then I was running into Glitch's max container size limit when trying `create-react-app` and other webpack+babel setups. After all that I was happy just to get something running.

I've now got correct sourcemaps set up for both demo sites. I can't re-run the traces right now, but if you re-run the traces you'll see accurate maps. I did a quick run, and `getImageData` is from `mapbox-gl-js`

<img width="1055" alt="screen shot 2017-10-23 at 7 55 44 pm" src="https://user-images.githubusercontent.com/57655/31922915-a0b827a2-b82c-11e7-9d45-9b61d4d79e46.png">

As are the long frames due to `Program` and `Texture.update` I mentioned
<img width="1154" alt="screen shot 2017-10-23 at 8 04 01 pm" src="https://user-images.githubusercontent.com/57655/31923053-71531ade-b82d-11e7-81d6-2c6daec269fd.png">

I'll give the `<AntiWobbleLayer>` approach a shot and report back.

**Comment by ilyabo** [comment]
I think I am actually having the same issue in my app (and using AntiWobbleLayer doesn't help with it).

**Comment by jfsiii** [comment]
@ibgreen I took a crack at using the `<AntiWobbleLayer>` [demo](https://loving-elk.glitch.me) [code](https://glitch.com/edit/#!/loving-elk?path=client/App.js) but it didn't change anything. I'm not sure I'm using it as intended though.

Since the issues appear to be inside `mapbox-gl-js`, I decided to make a demo using just that library; no React 

3MB GeoJSON [demo](https://tough-pest.glitch.me/) [code](https://glitch.com/edit/#!/tough-pes?path=client/App.js)
<img width="976" alt="screen shot 2017-10-25 at 4 40 35 pm" src="https://user-images.githubusercontent.com/57655/32028396-4b2459dc-b9a3-11e7-82ff-396efafd5bde.png">

30K GeoJSON [demo](https://south-condition.glitch.me/) [code](https://glitch.com/edit/#!/south-condition?path=client/App.js)
<img width="976" alt="screen shot 2017-10-25 at 4 39 54 pm" src="https://user-images.githubusercontent.com/57655/32028373-2ae46554-b9a3-11e7-8c60-61452d4f7344.png">

There are still long frames due to `getImageData` and `getProgramParameter`, but they're about 100-150ms for `getProgramParameter` and 250-350ms for `getImageData`. That's not bad.

I'll upload the traces soon and update this post.

I can also try to create an example using `react-map-gl` if you think it's worth it.

**Comment by samccone** [comment]
@jfsiii looks like your demos are no longer loading? glitch is returning 504s

**Comment by jfsiii** [comment]
@samccone :/ 

https://silver-saxophone.glitch.me/ is back up (the one with the 3MB GeoJSON file) but the other (https://reminiscent-glue.glitch.me) is still down. As I wrote https://github.com/uber/deck.gl/issues/1067#issuecomment-338859167 hosting this demo has been quite a struggle. I'll ping when it's back up and/or convert them into standalone repos.

The only difference between the two is the data source. The 3MB one is at https://gist.githubusercontent.com/jfsiii/b031dbbe7385d2c6822468a001d62225/raw/8ad146ac7db178b82480b8eb46eaaf406f60dc5d/USA-boundaries.json

and the 30K one is at https://gist.githubusercontent.com/jfsiii/381ea45844291c1d142e47f071f38396/raw/93b934b4f994c1d19a38a459ea2a15183eed8273/USA-boundaries-simple.json

**Comment by georgios-uber** [comment]
@jfsiii are you using fp64? Can you set fp64 to false and try again?

**Comment by garamond** [comment]
I have made a related observation: I added the deck.gl layers after the MapGL component finished rendering (with `setTimeout`) – this resulted in a _freeze of the entire browser_ (Chrome 64.0.3262.0, MacBook Pro). 

I stopped the profiler _immediately_ after the layers became visible. Interestingly, the rendered layers show up in the dev tool [timeline](https://chromedevtools.github.io/timeline-viewer/?loadTimelineFromURL=https://gist.githubusercontent.com/garamond/3c64bd376844f710e8ca2a92480a70ba/raw/abf83f9a78512f541deedd68d8d1d93c4a76ee3c/deckgl.json) _right away_, while they appear on screen only after the freeze.

![image](https://user-images.githubusercontent.com/299766/32662428-d0136b0e-c62a-11e7-9595-cdc447e45ef1.png)

**Comment by ilyabo** [comment]
We also experience the freeze during the layer initialization. The freeze goes away when deck.gl layers are removed or passed empty arrays as data. 

Notably the freeze is much more pronounced in Chrome on Windows. It can take 30 seconds to initialize deck.gl layers even for small datasets.

**Comment by shaojingli** [comment]
@garamond @ilyabo Hi guys, are you using fp64 flag in your layers?

**Comment by ilyabo** [comment]
Yes

**Comment by 1chandu** [comment]
@ilyabo we recently fixed a performance issue that might be related, it required changes in both luma.gl and deck.gl :
https://github.com/uber/luma.gl/pull/352
https://github.com/uber/deck.gl/pull/1147

We will soon post patch releases, above changes are in master, if you can please try with latest master of luma.gl and deck.gl

**Comment by ilyabo** [comment]
I tried to use our app with the latest luma.gl and deck.gl built from the master branches. It doesn't appear to resolve the problem with the freeze. 

Actually, the freeze can be reproduced with the layer browser app from deck.gl and it is very pronounced on Windows. I made this screen recording to illustrate the problem: https://www.youtube.com/watch?v=AGVnZZtA0Tk&feature=youtu.be   

You can see that Chrome is completely frozen from 0:05 to 0:35 after selecting GeoJSONLayer. Notably, after this initial freeze clicking on other layers doesn't lead to further freezes.

**Comment by ibgreen** [comment]
@ilyabo luma.gl and deck.gl 5.0 are now officially released, with very extensive under the hood changes, several of which affect performance. 
* So, it could be worth retesting on 5.0. 
* Also, we saw significant improvements for several other perf issues (on Mac) when Chrome 62 released, so good to measure on that version.

A current track of investigation is that we have traced certain excessive delays during layer startup to WebGL program linking. 
* We noticed that some GPU drivers seem to do the bulk of compilation/unrolling during program linking rather than during shader compilation.
* The linking delay particularly affects the large fp64 shaders and seems much more pronounced on Windows than Mac (reports of up to 30s in some cases)
* We already cache and reuse compiled WebGL shaders but that does not seem to make a difference. 
* It is much harder to cache and reuse linked Programs since each WebGL program instance contains all the uniform and attribute state. And even so we need to link at least once.
*We noticed that small changes to the shader code can have dramatic effects on link times (in one case a few lines made a 2x difference) but we don't yet have a strong intuition for what type of GLSL constructs to avoid.

**Comment by ilyabo** [comment]
Thanks for the status update and great that you are looking at this! I was trying with the current deck.gl master branch yesterday and could still see the freeze. I noticed also that setting fp64 to false eliminates the freeze. Will be testing further with 5.0.

**Comment by jianhuang01** [comment]
Tested the example, didn't find any difference with 64 enabled or not, both are around 400 ms with deck.gl 4.1.x, 1.0 s with 5.0.0. video card is Intel HD Graphics 630 1536 MB.

**Comment by ilyabo** [comment]
This is weird, we have experienced it on multiple machines, but maybe not with such good graphic cards. Did you try it on Windows?

**Comment by ilyabo** [comment]
And it still remains in 5.0, so we actually decided to rewrite it overlays in fp32

**Comment by ilyabo** [comment]
Check the YouTube video I posted above with the layer browser from the library. The freeze is 30sec on Windows

**Comment by Pessimistress** [comment]
Closing because this issue is referring to a version that is no longer maintained. Please open a new issue if you still experience this in the latest release.
